framework/include/userobject/ElementQualityChecker.h

```diff
@@ -0,0 +1,58 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef ELEMENTQUALITYCHECKER_H
+#define ELEMENTQUALITYCHECKER_H
+
+#include "ElementUserObject.h"
+
+#include "libmesh/enum_elem_quality.h"
+
+class ElementQualityChecker;
+
+template <>
+InputParameters validParams<ElementQualityChecker>();
+
+class ElementQualityChecker : public ElementUserObject
+{
+public:
+  ElementQualityChecker(const InputParameters & parameters);
+
+  static MooseEnum QualityMetricType();
+  static MooseEnum FailureMessageType();
+
+  void initialize() override;
+  void execute() override;
+  void threadJoin(const UserObject & uo) override;
+  void finalize() override;
+
+protected:
+  bool checkMetricApplicability(ElemQuality elem_metric, std::vector<ElemQuality> elem_metrics);
+
+  ElemQuality _m_type;
+  const bool _has_upper_bound;
+  const bool _has_lower_bound;
+  Real _upper_bound;
+  Real _lower_bound;
+
+  // set to save quality metric value for all elements
+  std::set<Real> _m_values;
```

Comment:
Bad idea. Using Reals for keys causes lots of problems. The problem is the comparison operator when inserting values that are close to one and other.

Reply:
I agree.  We just talked, and it doesn't need to be done this way.